### PR TITLE
Thapaan-CodeReview-2

### PR DESCRIPTION
**Pull request by ThapaArchana**
- Analysis of the program
 FindMyRecycling is an MVVM framework-based app that allows users to find recycling plants in their area and save them for later viewing.

- Was the program available in UC Github on time?
Program was available on GitHub on time

-Is the program documented/commented well enough for you to understand?
Yes the program is well documented and commented.

- Does the program compile?
 Yes 

- Rationale behind your changes.
1. Setting up the internet permission at manifest level to be able to get data through API using Retrofit with Kotlin Coroutine.
2. Implementation of Interface to remove concretion.
3. Autocomplete feature for fewer keystrokes and errors.

Links to my commits:

[Commit 1](https://github.com/phanga/MyPerishablePlannerV21001/commit/4781ce91953ef105dcd073067b2bb3883d8e5579)
[Commit 2](https://github.com/phanga/MyPerishablePlannerV21001/commit/d4b55111ab2f4269d303f35b262260c1cf21c0aa)
[Commit 3](https://github.com/phanga/MyPerishablePlannerV21001/commit/f71a608b1ad9335b9ab42245f390783d7f5ddb27)

- Three specific technical concepts that I learned from this code review

1. Using a Retrofit annotation to hit an API endpoint to retrieve JSON data
2. Implementation of Interface to inherit the behavior
3. State and composition

### CODE REVIEW

File: app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -1,8 +1,10 @@
 <?xml version="1.0" encoding="utf-8"?>
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
     package="com.findmyrecycling">
+    <uses-permission android:name="android.permission.INTERNET"/>
```
Comment (ThapaArchana) [inline app/src/main/AndroidManifest.xml:4]: Add Internet permission in the manifests file.
https://www.geeksforgeeks.org/retrofit-with-kotlin-coroutine-in-android/

File: app/src/main/java/com/findmyrecycling/service/ProductService.kt [hunk 2]
```diff
@@ -8,13 +9,17 @@ import kotlinx.coroutines.async
...
+class ProductService : IProductService {
+    override suspend fun fetchProducts() : List <Product>? {
         return withContext(Dispatchers.IO) {
             val service = RetrofitClientInstance.retrofitInstance?.create(IProductDAO::class.java)
-            val products = async {service?.getAllProducts()}
-            var result = products.await()?.awaitResponse()?.body()
+            val items = async { service?.getAllProducts()}
+            var result = items.await()?.awaitResponse()?.body()
             return@withContext result
         }
```
Comment (ThapaArchana) [inline app/src/main/java/com/findmyrecycling/service/ProductService.kt:24]: Implemented interface to remove concretion.
https://oozou.com/blog/interface-in-kotlin-and-when-to-use-it-10

File: app/src/main/java/com/findmyrecycling/MainActivity.kt [hunk 3]
```diff
@@ -22,91 +19,186 @@ import com.firebase.ui.auth.AuthUI
...
+    fun TextFieldWithDropdownUsage(productsIn: List<Product>, label: String = "", take: Int= 3 ) {
+
+        val dropDownOptions = remember { mutableStateOf(listOf<Product>()) }
+        val textFieldValue = remember { mutableStateOf(TextFieldValue()) }
+        val dropDownExpanded = remember { mutableStateOf(false) }
+
+        fun onDropdownDismissRequest() {
+            dropDownExpanded.value = false
+        }
+
```
Comment (ThapaArchana) [inline app/src/main/java/com/findmyrecycling/MainActivity.kt:131]: Added Autocomplete for fewer keystrokes and correcting user query and spelling errors.
https://blog.expertrec.com/autocomplete/

**Comment by discospiff** [comment]
It looks like new functionality is added in this code review (Koin, Auto Complete), and those features appear to be complete.

I encourage the team to review these changes, and if they are appropriate for the product backlog, merge this branch.

**Pull request closed without merging**